docs/dev/explanations/tasks-architecture.md

```diff
@@ -0,0 +1,124 @@
+- **Object level locking:** Tasks can also optionally claim ownership on a DiracX resource (e.g. job or transformation). Several types of lock are available: mutex, semaphore and readers–writer. Additionally, the failure mode of locks can be configured to either reset with a time-to-live or require manual recovery.
+
+### Retries
+
+Two forms of retry are available to tasks:
+
+- **Reschedule retries:** Pre-requisites are not satisfied, these retries can be configured to repeat without hard limits.
+- **Failure retries:** The task failed for an unknown reason that may be ephemeral. A configurable hard upper limit is placed on the number of failure retries.
+
+When retrying each type of failure has a configurable policy of when the next attempt is applied ranging from immediate retries to various backoff strategies.
+
+## Broker
+
+The state of the broker should be ephemeral and recreated with each update. Any persistent state should be stored in the standard MySQL database's used by DiracX. This requirement is imposed to:
+
+- Simplify recovery from unexpected outages.
+- Reduce the complexity of reasoning about updates which may change details of the broker's internal state.
+- Improve performance by removing the need to ensure every action is flushed to persistent storage.
+
+Upon first start, the broker is populated with the cron-style tasks as well as any pending reactive tasks that have been persisted in MySQL. The tasks which were persisted are those eligible for the dead letter queue.
+
+## Operational considerations
+
+### Configuration
+
+The configuration of tasks can be done through a dedicated DiracX configuration file. This file includes per-task overrides for:
+
+- Cron-style schedules
+- Rate limits
+- Number of retry attempts
+- Backoff behaviour
+
+Default values for the above are included in the task definitions in the DiracX source code.
+
+### Monitoring
+
+All tasks should keep track of:
+
+- When they were first submitted.
+- How many times they have failed.
+- How many times they have attempted to be executed. This may be different from the number of failures if the task is configured to defer itself based on rate limits or the status of an external resource.
+- The parent task which caused them to be called (if any).
+- The resource consumption of their execution.
+- The number of failed attempts and debugging information about each failed attempt.
+
+The broker should provide functionality to:
+
+- Understand the size and composition of pending and deferred tasks.
+- Observe the currently running tasks.
+
+### Resource utilisation
+
+To ensure the stability of the system, all workers should be configured to enforce memory and CPU limits. If these limits are exceeded the DiracX task system must be able to detect and report the issue, however in the case of small workers such detection may be unreliable due to technical constraints in async environments.
```

Comment:
Are workers organized into separate pools based on resource capacity?
Or do they have identical resource allocation sized for large tasks?

How do we plan to connect task queues to workers?